src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.SingleTC.cs

```diff
@@ -0,0 +1,104 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    using static BinaryOperatorKind;
+
+    internal static partial class ValueSetFactory
+    {
+        private struct SingleTC : FloatingTC<float>, NumericTC<float>
+        {
+            float NumericTC<float>.MinValue => float.MinValue;
+
+            float NumericTC<float>.MaxValue => float.MaxValue;
+
+            float FloatingTC<float>.NaN => float.NaN;
+
+            float FloatingTC<float>.MinusInf => float.NegativeInfinity;
+
+            float FloatingTC<float>.PlusInf => float.PositiveInfinity;
+
+            float NumericTC<float>.Next(float value)
+            {
+                Debug.Assert(!float.IsNaN(value));
+                Debug.Assert(!float.IsInfinity(value));
+                Debug.Assert(value != float.MaxValue);
+                if (value < 0)
+                {
+                    if (value == -float.Epsilon)
+                        return 0.0f; // skip negative zero
+                    return -UintAsFloat(FloatAsUint(-value) - 1);
+                }
+                return UintAsFloat(FloatAsUint(value) + 1);
+            }
+
+            private unsafe static uint FloatAsUint(float d)
+            {
+                float* dp = &d;
+                uint* lp = (uint*)dp;
+                return *lp;
+            }
+
+            private unsafe static float UintAsFloat(uint l)
+            {
+                uint* lp = &l;
+                float* dp = (float*)lp;
+                return *dp;
+            }
+
+            (float leftMax, float rightMin) NumericTC<float>.Partition(float min, float max)
+            {
+                Debug.Assert(min != max);
+
+                if (min == float.MinValue && max == float.MaxValue)
+                    return (-UintAsFloat(1), 0.0f); // skip negative zero
+
+                Debug.Assert((min >= 0) == (max >= 0));
+                if (min < 0)
+                {
+                    uint minl = FloatAsUint(-max);
+                    uint maxl = FloatAsUint(-min);
+                    uint midl = minl + (maxl - minl) / 2;
+                    return (-UintAsFloat(midl + 1), -UintAsFloat(midl));
+                }
+                else
+                {
+                    uint minl = FloatAsUint(min);
+                    uint maxl = FloatAsUint(max);
+                    uint midl = minl + (maxl - minl) / 2;
+                    return (UintAsFloat(midl), UintAsFloat(midl + 1));
+                }
+            }
+
+            bool NumericTC<float>.Related(BinaryOperatorKind relation, float left, float right)
+            {
+                switch (relation)
+                {
+                    case Equal:
+                        return left == right || float.IsNaN(left) && float.IsNaN(right); // for our purposes, NaNs are equal
```

Comment:
>// for our purposes, NaNs are equal [](start = 89, length = 35)

Can we add a C# test for NaN?

Reply:
We already have that.  It is an error.  See `ERR_RelationalPatternWithNaN` and tests that produce it like `Relational_06`

---
In reply to: [381445814](https://github.com/dotnet/roslyn/pull/41503#discussion_r381445814) [](ancestors = 381445814)